File: services/api/app/schemas/notifications.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class NotificationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    user_id: UUID
    workspace_id: UUID
    type: str
    title: str
    body: str | None
    severity: str | None
    link_kind: str | None
    link_id: UUID | None
    read_at: datetime | None
    created_at: datetime


class NotificationList(BaseModel):
    items: list[NotificationOut]
    next_cursor: str | None
    unread_count: int

File: apps/web/lib/notifications.ts
import { api } from './api';

export type Notification = {
  id: string;
  user_id: string;
  workspace_id: string;
  type: 'insight_created' | 'ingest_completed' | 'ingest_failed';
  title: string;
  body: string | null;
  severity: 'low' | 'medium' | 'high' | 'info' | null;
  link_kind: 'insight' | 'document' | null;
  link_id: string | null;
  read_at: string | null;
  created_at: string;
};

export type NotificationList = {
  items: Notification[];
  next_cursor: string | null;
  unread_count: number;
};

export const notificationsApi = {
  list: (params?: { unread?: boolean; limit?: number }) => {
    const qs = new URLSearchParams();
    if (params?.unread !== undefined) qs.set('unread', String(params.unread));
    if (params?.limit) qs.set('limit', String(params.limit));
    const suffix = qs.toString() ? `?${qs}` : '';
    return api<NotificationList>(`/api/notifications${suffix}`);
  },
  markRead: (id: string) =>
    api<Notification>(`/api/notifications/${id}`, { method: 'PATCH' }),
  markAllRead: () =>
    api<void>('/api/notifications/mark-all-read', { method: 'POST' }),
};
